.github/workflows/ci.yml

```diff
@@ -15,6 +15,10 @@ env:
   TEST_TIMEOUT: 360
   MACOS_ONEAPI_DOWNLOAD_LINK: https://registrationcenter-download.intel.com/akdlm/irc_nas/18358/m_cpp-compiler-classic_p_2022.0.0.62_offline.dmg
 
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
+  cancel-in-progress: true
+
```

Comment:
Will it also cancel runs for the main branch? If so, it is better add the main branch as an exception.

Reply:
I believe since it references the `github.event.pull_request.number` it should not touch the workflows on the main branch.  I have not tested this though.

Reply:
> Will it also cancel runs for the main branch? If so, it is better add the main branch as an exception.

Why it's needed?

Reply:
Oh actually, I understand the question.  Yes this may be a problem now that I think about it.  Thanks for bringing it up.

If two PRs are merged in quick succession, we don't want the second merged to cancel the first's run.  Let me look into this.  I'll try to find documentation about what those variables are for main branch, and how we might make an exception here.

Reply:
You mean we should have one build for each commit in the `main` branch?

Reply:
We don't want to cancel post-merge CI runs on main.  @dmitriy-sobolev  provided a solution: https://stackoverflow.com/questions/68418857/how-to-cancel-existing-runs-when-a-new-push-happens-on-github-actions-but-only/70972844#70972844

Reply:
> > Will it also cancel runs for the main branch? If so, it is better add the main branch as an exception.
> 
> @dmitriy-sobolev Why it's needed?

I do not have strong reasoning, but `main` branch has more "coarse-grained" commits compared to feature branches, it also should not have "intermediate" states, which we can skip. It will also facilitate finding regressions if any.